internal/parse/parser.go

```diff
@@ -1,57 +1,140 @@
 package parse
 
-import "strings"
+import (
+	"strings"
+)
 
-// Parser is used to parse the SQLAir DSL.
 type Parser struct {
-	// input is the DSL statement to be parted.
 	input string
-
-	// lastParsedPos is the character position of the last parsed part.
-	lastParsedPos int
-
-	// pos is the current character position of the parser.
-	pos int
+	pos   int
 }
 
-// NewParser returns a reference to a new parser.
 func NewParser() *Parser {
 	return &Parser{}
 }
 
-// init initializes the parser.
+// init resets the state of the parser and sets the input string.
 func (p *Parser) init(input string) {
 	p.input = input
-	p.lastParsedPos = 0
 	p.pos = 0
 }
 
-func (p *Parser) Parse(input string) (*ParsedExpr, error) {
-	p.init(input)
-	return nil, nil
+// A checkpoint struct for saving parser state to restore later. We only use
+// a checkpoint within an attempted parsing of an part, not at a higher level
+// since we don't keep track of the parts in the checkpoint.
+type checkpoint struct {
+	parser *Parser
+	pos    int
+}
+
+// save takes a snapshot of the state of the parser and returns a pointer to a
+// checkpoint that represents it.
+func (p *Parser) save() *checkpoint {
+	return &checkpoint{
+		parser: p,
+		pos:    p.pos,
+	}
+}
+
+// restore sets the internal state of the parser to the values stored in the
+// checkpoint.
+func (cp *checkpoint) restore() {
+	cp.parser.pos = cp.pos
 }
 
-// ParsedExpr represents a parsed expression.
-// It has a representation of the original SQL statement in terms of QueryParts
+// advance moves the parser's index forward by one element.
+func (p *Parser) advance() bool {
+	p.skipSpaces()
+	if p.pos == len(p.input) {
+		return false
+	}
+	p.pos++
+	return true
+}
+
+// ParsedExpr is the AST representation of an SQL expression.
+// It has a representation of the original SQL statement in terms of queryParts
 // A SQL statement like this:
 //
 // Select p.* as &Person.* from person where p.name = $Boss.Name
 //
 // would be represented as:
 //
-// [stringPart outputPart stringPart inputPart]
+// [BypassPart OutputPart BypassPart InputPart]
 type ParsedExpr struct {
 	queryParts []queryPart
 }
 
-// peekByte returns true if the current byte
-// equals the one passed as parameter.
+// String returns a textual representation of the AST contained in the
+// ParsedExpr for debugging purposes.
+func (pe *ParsedExpr) String() string {
+	out := "ParsedExpr["
+	for i, p := range pe.queryParts {
+		if i > 0 {
+			out = out + " "
+		}
+		out = out + p.String()
+	}
+	out = out + "]"
+	return out
+}
+
+// parsedExprBuilder keeps track of the parts parsed so far, along with
+// information for parsing the BypassPart between input/output parts.
+type parsedExprBuilder struct {
```

Comment:
It's also not clear what the benefit of this construct is, when we already have a parser that is already re-initialized at the same moment that this type is created. We might make a Parser that is stateless and holds the state in memory until the parsing is done, and then it might be safely used in parallel without races. But as it is we have neither, as half of the state of the parsing is in Parse, and half is in the builder.  Am I missing something?  Don't we get to the same place if these fields are in Parser itself?

Reply:
The goal is noble, but per comment above the problem is that we cannot claim there is any clarity improvement for this particular case, because the Parser and the ParserExprBuilder are strictly defined in terms of each other's implementations. The Parser has some state which is re-initialized every time a parsing happens, and at that exact moment a new Builder is created. In other words, the Parser state and the Builder state are strictly connected and reset at once, and the Parser knows about the Builder's details, while the Builder knows about the exact implementation of the Parser and its private fields. So these don't really look like two different types.. they are indeed a single type with two names, with no clear abstraction or encapsulation across them.

We can clean this up in at least two ways:

1. Take the rest of parsing logic content outside of Parser itself. The only benefit I can see of doing this is that Parser becomes stateless, but then why even have a type for this?

2. Otherwise, merge the Builder back into the Parser itself, which also makes some sense as the "parts builder" is really the only point of having a Parser at all (it parses the expression).